### PR TITLE
SNOW-896744: Allow creating dataframes using tuples as schema

### LINKED ISSUE #1009: SNOW-896744: creating a DataFrame with a tuple of names will be ignored an generic names will be used
Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using?

 Python 3.10.11 (v3.10.11:7d4cc5aa85, Apr  4 2023, 19:05:19) [Clang 13.0.0 (clang-1300.0.29.30)]

2. What operating system and processor architecture are you using?

macOS-13.5-arm64-arm-64bit

3. What are the component versions in the environment (`pip freeze`)?

absl-py==1.4.0
aiofiles==22.1.0
aiohttp==3.8.4
aiosignal==1.3.1
aiosqlite==0.18.0
altair==5.0.1
altgraph @ file:///System/Volumes/Data/SWE/Apps/DT/BuildRoots/BuildRoot7/ActiveBuildRoot/Library/Caches/com.apple.xbs/Sources/python3/python3-133.100.1.1/altgraph-0.17.2-py2.py3-none-any.whl
anthropic @ file:///Users/mrojas/llm-research/ANTROPIC
anyio==3.6.2
anytree==2.8.0
appnope==0.1.3
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
arrow==1.2.3
asn1crypto==1.5.1
ast2json==0.3
astor==0.8.1
asttokens==2.2.1
async-timeout==4.0.2
attrs==22.2.0
autograd==1.5
autograd-gamma==0.5.0
aws-secretsmanager-caching==1.1.1.5
Babel==2.11.0
backcall==0.2.0
beautifulsoup4==4.11.2
bleach==6.0.0
blinker==1.6.2
boto3==1.28.14
botocore==1.31.14
bs4==0.0.1
build==0.10.0
bytecode==0.14.1
CacheControl==0.13.1
cachetools==5.3.0
CD==1.1.0
certifi==2023.7.22
cffi==1.15.1
cfgv==3.3.1
charset-normalizer==3.2.0
cleo==2.0.1
click==8.1.6
cloudpickle==2.0.0
cobble==0.1.3
colorama==0.4.6
comm==0.1.2
contourpy==1.0.7
coverage==7.2.7
crashtest==0.4.1
cryptography==41.0.3
csv-schema-inference==0.0.9
cycler==0.11.0
cyclonedx-python-lib==4.0.1
cypari==2.4.1
databricks-cli==0.17.7
DateTime==5.2
debugpy==1.6.6
decorator==5.1.1
defusedxml==0.7.1
Deprecated==1.2.13
distlib==0.3.7
dukpy==0.3.0
dulwich==0.21.5
editables==0.3
et-xmlfile==1.1.0
exceptiongroup==1.1.1
executing==1.2.0
extra-streamlit-components==0.1.56
fastjsonschema==2.16.2
filelock==3.12.2
findspark==2.0.1
Flask==2.2.3
fonttools==4.39.3
format-sql==0.12.0
formulaic==0.6.1
fqdn==1.5.1
frozenlist==1.3.3
fsspec==2023.1.0
future @ file:///System/Volumes/Data/SWE/Apps/DT/BuildRoots/BuildRoot7/ActiveBuildRoot/Library/Caches/com.apple.xbs/Sources/python3/python3-133.100.1.1/future-0.18.2-py3-none-any.whl
FXrays==1.3.5
gitdb==4.0.10
GitPython==3.1.32
google-auth==2.22.0
google-auth-oauthlib==1.0.0
graphviz==0.20.1
gspread==5.10.0
h11==0.14.0
hatch==1.7.0
hatchling==1.14.1
html2markdown==0.1.7
html5lib==1.1
httpcore==0.16.3
httpx==0.23.3
hyperlink==21.0.0
identify==2.5.19
idna==3.4
importlib-metadata==6.8.0
importlib-resources==5.12.0
iniconfig==2.0.0
installer==0.7.0
interface-meta==1.3.0
ipykernel==6.21.2
ipython==8.10.0
ipython-genutils==0.2.0
ipywidgets==8.0.4
isoduration==20.11.0
itsdangerous==2.1.2
jaraco.classes==3.3.0
javascripthon==0.12
jedi==0.18.2
Jinja2==3.1.2
jiphy==1.2.2
jmespath==1.0.1
joblib==1.2.0
johnnydep==1.17.5
json5==0.9.11
jsonpointer==2.3
jsonschema==4.17.3
jupyter==1.0.0
jupyter-console==6.5.1
jupyter-events==0.6.3
jupyter-ydoc==0.2.2
jupyter_client==8.0.3
jupyter_core==5.2.0
jupyter_server==2.3.0
jupyter_server_fileid==0.6.0
jupyter_server_terminals==0.4.4
jupyter_server_ydoc==0.6.1
jupyterlab==3.6.1
jupyterlab-pygments==0.2.2
jupyterlab-widgets==3.0.5
jupyterlab_server==2.19.0
keyring==23.13.1
kiwisolver==1.4.4
knot-floer-homology==1.2
lifelines==0.27.7
llvmlite==0.40.1
lockfile==0.12.2
lxml==4.9.3
macholib @ file:///System/Volumes/Data/SWE/Apps/DT/BuildRoots/BuildRoot7/ActiveBuildRoot/Library/Caches/com.apple.xbs/Sources/python3/python3-133.100.1.1/macholib-1.15.2-py2.py3-none-any.whl
mammoth==1.5.0
markdown-it-py==3.0.0
markdownify==0.11.6
markov-clustering==0.0.6.dev0
MarkupSafe==2.1.3
matplotlib==3.7.1
matplotlib-inline==0.1.6
mdurl==0.1.2
mistune==2.0.5
more-itertools==10.0.0
msgpack==1.0.5
multidict==6.0.4
nbclassic==0.5.1
nbclient==0.7.2
nbconvert==7.2.9
nbformat==5.7.3
nest-asyncio==1.5.6
networkx==3.0
nodeenv==1.7.0
notebook==6.5.2
notebook_shim==0.2.2
numba==0.57.1
numpy==1.24.4
nyoka==5.5.0
oauthlib==3.2.2
openpyxl==3.1.2
oscrypto==1.3.0
outcome==1.2.0
oyaml==1.0
packageurl-python==0.11.2
packaging==23.1
pandas==2.0.3
pandocfilters==1.5.0
parquet==1.3.1
parso==0.8.3
pathspec==0.11.1
pdfminer==20191125
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.5.0
pip-api==0.0.30
pip-requirements-parser==32.0.1
pip_audit==2.6.1
pkginfo==1.9.6
platformdirs==3.8.1
plink==2.4.1
plotly==5.15.0
pluggy==1.0.0
ply==3.11
poetry==1.5.1
poetry-core==1.6.1
poetry-plugin-export==1.4.0
pre-commit==3.1.1
prometheus-client==0.16.0
prompt-toolkit==3.0.36
protobuf==4.23.4
psutil==5.9.4
ptyprocess==0.7.0
pure-eval==0.2.2
py-serializable==0.11.1
py-snappy==0.1.0a1
py4j==0.10.9.5
pyarrow==10.0.1
pyasn1==0.5.0
pyasn1-modules==0.3.0
pycparser==2.21
pycryptodome==3.17
pycryptodomex==3.18.0
pydeck==0.8.0
pydot==1.4.2
PyGithub==1.58.2
Pygments==2.15.1
PyJWT==2.8.0
Pympler==1.0.1
PyNaCl==1.5.0
pyngrok==6.0.0
pyOpenSSL==23.2.0
pyparsing==3.0.9
pyperclip==1.8.2
pypng==0.20220715.0
pyproject_hooks==1.0.0
pyrsistent==0.19.3
PySocks==1.7.1
pyspark==3.3.2
pytest==7.3.1
python-dateutil==2.8.2
python-decouple==3.8
python-gitlab==3.15.0
python-json-logger==2.0.6
python-snappy==0.6.1
pytz==2023.3
pytz-deprecation-shim==0.1.0.post0
pyxlsb==1.0.10
PyYAML==6.0.1
pyzmq==25.0.0
qtconsole==5.4.0
QtPy==2.3.0
rapidfuzz==2.15.1
requests==2.31.0
requests-oauthlib==1.3.1
requests-toolbelt==1.0.0
requirements-parser==0.5.0
rfc3339-validator==0.1.4
rfc3986==1.5.0
rfc3986-validator==0.1.1
rich==13.4.2
rsa==4.9
s3transfer==0.6.1
scikit-learn==1.2.1
scipy==1.10.1
selenium==4.10.0
Send2Trash==1.8.0
shellingham==1.5.0.post1
simplejson==3.19.1
six==1.16.0
smmap==5.0.0
snappy==3.0.3
snappy-manifolds==1.1.2
sniffio==1.3.0
snowconvert-deploy-tool==0.0.25
snowflake-connector-python==3.1.0
snowflake-connector-python-nightly==2023.6.24
snowflake-ml-python @ file:///Users/mrojas/Downloads/snowflake_ml_python-1.0.2-py3-none-any.whl
snowflake-snowpark-python==1.6.1
-e git+https://github.com/Snowflake-Labs/snowpark-extensions-py.git@ec4a706f6709abe25955212b7b9d7e3fa663f959#egg=snowpark_extensions
sortedcontainers==2.4.0
soupsieve==2.4
spherogram==2.1
sql-formatter==0.6.2
sql-metadata==2.8.0
SQLAlchemy==2.0.15
sqlparse==0.4.4
stack-data==0.6.2
streamlit==1.25.0
streamlit-aggrid==0.3.4.post3
streamlit-option-menu==0.3.6
structlog==22.3.0
tabulate==0.9.0
tenacity==8.2.2
terminado==0.17.1
threadpoolctl==3.1.0
thriftpy2==0.4.16
tinycss2==1.2.1
tokenizers==0.13.2
toml==0.10.2
tomli==2.0.1
tomli_w==1.0.0
tomlkit==0.12.1
toolz==0.12.0
tornado==6.2
traitlets==5.9.0
trio==0.22.1
trio-websocket==0.10.3
trove-classifiers==2023.4.25
typer==0.9.0
types-setuptools==68.0.0.3
typing_extensions==4.7.1
tzdata==2023.3
tzlocal==4.3.1
uri-template==1.2.0
urllib3==1.26.16
userpath==1.8.0
validators==0.20.0
virtualenv==20.24.2
wcwidth==0.2.6
webcolors==1.12
webencodings==0.5.1
websocket-client==1.5.1
Werkzeug==2.2.3
widgetsnbextension==4.0.5
wimpy==0.6
wrapt==1.15.0
wsproto==1.2.0
xattr==0.10.1
xgboost==1.7.6
XlsxWriter==3.1.2
y-py==0.5.5
yarl==1.9.2
ypy-websocket==0.8.2
zipp==3.16.2
zope.interface==6.0

4. What did you do?

df2 = session.createDataFrame(
    [(20000101, 1, "x"), (20000101, 2, "y")],
    schema=("TIME", "ID", "V2"))

5. What did you expect to see?

  If I do df2.show() it should be:

--------------------------
|"TIME"    |"ID"  |"V1"  |
--------------------------
|20000101  |1     |1.0   |
|20000101  |2     |2.0   |
|20000102  |1     |3.0   |
|20000102  |2     |4.0   |
--------------------------

But currently it shows 
--------------------------
|"_1"      |"_2"  |"_3"  |

This works for when I use a list of strings. 
I think this should work with any iterable

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #1009 

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/session.py [hunk 1]
```diff
@@ -1865,7 +1865,7 @@ def write_pandas(
     def create_dataframe(
         self,
         data: Union[List, Tuple, "pandas.DataFrame"],
-        schema: Optional[Union[StructType, List[str]]] = None,
+        schema: Optional[Union[StructType, List[str], Tuple[str]]] = None,
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/session.py:1868]: Should we make sure we use `schema` as an iterable when it is not `StructType`, and then support `Iterable[str]`?

File: src/snowflake/snowpark/_internal/type_utils.py [hunk 2]
```diff
@@ -325,7 +325,7 @@ def infer_type(obj: Any) -> DataType:
 
 
 def infer_schema(
-    row: Union[Dict, List, Tuple], names: Optional[List] = None
+    row: Union[Dict, List, Tuple], names: Optional[Union[List, Tuple]] = None
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/type_utils.py:328]: I don't think we could do `names.extend`. How about we explicitly copy `names` into a list first? This way we could support `Iterable`.

**Review by sfc-gh-sfan — CHANGES_REQUESTED** [review]

**Comment by sfc-gh-sfan** [comment]
Did not mean to "requested changes".  The change looks good overall but wonder if we could extend to `Iterable`

File: src/snowflake/snowpark/_internal/type_utils.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/type_utils.py:328]: You are right. I can convert iterable into a list and then assume list datatype from thereon...

**Review by sfc-gh-sfan — APPROVED** [review]
